### PR TITLE
[IMP] edi: Define a processing function for incoming documents

### PR DESCRIPTION
**Pull request by etobella**
Adding the processing function.

@simahawk @jarroyomorales

### CODE REVIEW

File: edi/models/edi_backend.py [hunk 1]
```diff
@@ -127,3 +127,63 @@ def _validate_generate_output(self, exchange_record, force=False):
...
+    def _exchange_input_check(self, record):
+        return record.edi_exchange_state in ["input_received", "input_processed_error"]
+
+    def exchange_process(self, exchange_record):
+        """
+        This function should be called when an exchange record has been received
+        it could integrate check where to relate or modificate the data
+        """
+        self.ensure_one()
+        exchange_record.ensure_one()
```
Comment (simahawk) [inline edi/models/edi_backend.py]: same as per the output: move to _exchange_input_check

File: edi/models/edi_backend.py [hunk 2]
```diff
@@ -127,3 +127,63 @@ def _validate_generate_output(self, exchange_record, force=False):
...
+                "level": level,
+            },
+            subtype_id=self.env.ref("mail.mt_note").id,
+        )
+
+    def _exchange_input_check(self, record):
+        return record.edi_exchange_state in ["input_received", "input_processed_error"]
+
+    def exchange_process(self, exchange_record):
+        """
```
Comment (simahawk) [inline edi/models/edi_backend.py]: https://www.python.org/dev/peps/pep-0257/#multi-line-docstrings

File: edi/models/edi_backend.py [hunk 3]
```diff
@@ -127,3 +127,63 @@ def _validate_generate_output(self, exchange_record, force=False):
...
+            res = True
+        finally:
+            exchange_record.edi_exchange_state = state
+            exchange_record.exchange_error = error
+            if message:
+                self._exchange_notify_record(exchange_record, message)
+        return res
+
+    def _exchange_process(self, exchange_record):
+        # TODO: Maybe we could contact to the component here. Isn't it?
```
Comment (simahawk) [inline edi/models/edi_backend.py]: yeah something similar as done for output

File: edi/models/edi_backend.py [hunk 4]
```diff
@@ -179,11 +179,11 @@ def _exchange_send(self, exchange_record):
         raise NotImplementedError()
 
     def _exchange_notify_record(self, record, message, level="info"):
-        """Attach notification of exchange state to the original record."""
+        """Attach exported file to original record."""
         if not hasattr(record.record, "message_post_with_view"):
             return
         record.record.message_post_with_view(
-            "edi.message_edi_exchange_link",
+            "base_edi_exchange.message_edi_exchange_link",
```
Comment (simahawk) [inline edi/models/edi_backend.py]: @etobella bad copy/paste :wink:
```suggestion
            "edi.message_edi_exchange_link",
```

File: edi/models/edi_backend.py [hunk 5]
```diff
@@ -192,3 +192,59 @@ def _exchange_notify_record(self, record, message, level="info"):
...
+        self.ensure_one()
+        exchange_record.ensure_one()
+        # In case already processed: skip processing and check the state
+        check = self._exchange_input_check(exchange_record)
+        if not check:
+            return False
+        try:
+            self._exchange_process(exchange_record)
+        except Exception as err:
+            if self.env.context.get("_edi_send_break_on_error"):
```
Comment (simahawk) [inline edi/models/edi_backend.py]:
```suggestion
            if self.env.context.get("_edi_receive_break_on_error"):
```

**Review by simahawk — CHANGES_REQUESTED** [review]

**Comment by etobella** [comment]
Rebase and added the same logic

File: edi/models/edi_exchange_record.py [hunk 6]
```diff
@@ -147,3 +147,12 @@ def action_exchange_send(self):
         if not self.direction == "output":
             raise exceptions.UserError(_("An output record is required for sending!"))
         return self.backend_id.exchange_send(self)
+
+    def _exchange_processed_ok_msg(self):
```
Comment (simahawk) [inline edi/models/edi_exchange_record.py]: @etobella please check the new way to access this messages (no need for a new method every time)

File: edi/models/edi_backend.py [hunk 7]
```diff
@@ -229,3 +229,69 @@ def _exchange_notify_record(self, record, message, level="info"):
...
+        # In case already processed: skip processing and check the state
+        check = self._exchange_input_check(exchange_record)
+        if not check:
+            return False
+        try:
+            self._exchange_process(exchange_record)
+        except Exception as err:
+            if self.env.context.get("_edi_receive_break_on_error"):
+                raise
+            error = str(err)
```
Comment (simahawk) [inline edi/models/edi_backend.py]:
```suggestion
            error = repr(err)
```

File: edi/models/edi_backend.py [hunk 8]
```diff
@@ -229,3 +229,69 @@ def _exchange_notify_record(self, record, message, level="info"):
...
+        """
+        self.ensure_one()
+        exchange_record.ensure_one()
+        # In case already processed: skip processing and check the state
+        check = self._exchange_input_check(exchange_record)
+        if not check:
+            return False
+        try:
+            self._exchange_process(exchange_record)
+        except Exception as err:
```
Comment (simahawk) [inline edi/models/edi_backend.py]:
```suggestion
        except self._swallable_exceptions() as err:
```

**Comment by simahawk** [comment]
@etobella split what we talked about here  https://github.com/OCA/edi/pull/266

**Comment by simahawk** [comment]
@etobella please rebase your PR here and then check https://github.com/tegin/edi/pull/1 which adds on top of this almost all the changes required. See the message regarding tests there.

**Comment by etobella** [comment]
Rebase and PR merged :wink: I had some issues with the testing locally, it seem the representation on my computed differs from Travis, not a big issue although :smile:

**Comment by etobella** [comment]
Now it should be :green_apple:

File: edi_exchange_template/tests/test_edi_backend_output.py [hunk 9]
```diff
@@ -4,10 +4,10 @@
 
 from lxml import etree
 
-from odoo.addons.edi.tests.common import EDIBackendCommonTestCase
+from odoo.addons.edi.tests.common import EDIBackendCommonComponentRegistryTestCase
 
 
-class TestEDIBackendOutputBase(EDIBackendCommonTestCase):
+class TestEDIBackendOutputBase(EDIBackendCommonComponentRegistryTestCase):
```
Comment (simahawk) [inline edi_exchange_template/tests/test_edi_backend_output.py]: this is not needed

File: edi_exchange_template/tests/test_edi_backend_output.py [hunk 10]
```diff
@@ -92,6 +92,12 @@ def _setup_records(cls):
         }
         cls.record2 = cls.backend.create_record("test_type_out2", vals)
 
+    @classmethod
+    def setUpClass(cls):
```
Comment (simahawk) [inline edi_exchange_template/tests/test_edi_backend_output.py]: same

File: edi_exchange_template/tests/test_edi_backend_output.py [hunk 11]
```diff
@@ -72,7 +72,7 @@ def _setup_records(cls):
         )
         cls.tmpl_out2 = model.create(
             {
-                "code": "edi.output.demo_backend.test_type_out2",
+                "code": "edi.output.output.demo_backend.test_type_out2",
```
Comment (simahawk) [inline edi_exchange_template/tests/test_edi_backend_output.py]: this should be `edi.output.generate.*` and `_get_output_template` should be fixed -> `self._get_component_usage_candidates(exchange_record, "output")` should use `"generate"`.
@etobella do you have tests failing here? It's all good for me locally :thinking:

File: edi/components/base.py [hunk 12]
```diff
@@ -17,34 +17,3 @@ def exchange_record(self):
     @property
     def backend(self):
         return self.work.backend
-
```
Comment (simahawk) [inline edi/components/base.py:20]: @etobella something is wrong w/ your rebase... :thinking:

File: edi_exchange_template/tests/test_edi_backend_output.py (hunk 11 above)
Comment (etobella) [reply edi_exchange_template/tests/test_edi_backend_output.py]: Yes, it is failing, with the latest changes, as output and input have been added to the base of the code name

**Review by simahawk — APPROVED** [review]

**Pull request merged**